Disable Delete/Restore menu items appropriately

AbeJellinek opened the pull request:
An item can only be deleted from the trash if it or one of its ancestors is `.deleted`. An item can only be restored from the trash if it is `.deleted`.

Fixes #2301.

dstillman commented on chrome/content/zotero/zoteroPane.js:
```diff
@@ -1893,6 +1909,21 @@ var ZoteroPane = new function()
+			return false;
+		}
+
+		return this.getSelectedItems().every(item => item.deleted);
```
This isn't quite right. If you select a parent item and a child item in the trash, "Restore to Library" should be enabled, even if one of them isn't technically deleted.

AbeJellinek replied:
I'd like to get the logic right here, but it depends on what we're going for - should we only show Restore if it will cause everything you have selected to disappear from the trash? If so, we also need to check whether any of the selected items has a descendant that is deleted and not selected, since that would mean that part of your selection will remain in the trash after Restore. If we only care whether it'll clear _some_ of the selected items from the trash, the test can be simpler.

dstillman replied:
- Deleted/selected parent, deleted/selected child, no other children → show Restore, undelete both
- Non-deleted/selected parent, deleted/selected child, no other children → show Restore, undelete child (because there's no reason not to)
- Non-deleted/selected parent, deleted/selected child, deleted/non-selected children → show Restore, undelete selected child, don't undelete non-selected children (because the intention is clear)
- Deleted/selected parent, deleted/non-selected child, non-deleted/non-selected child → show Restore, restore parent and deleted child (because there's no way to know the user's intention and this is safer)
- Deleted/selected parent, non-deleted/selected child, deleted/non-selected child → show Restore, restore parent only, don't undelete non-selected child (because selecting a specific child communicates intention)
- Non-deleted/selected parent, deleted/non-selected child → don't show Restore (because they didn't select anything non-gray)
- Non-deleted/non-selected parent, deleted/selected child, non-deleted/selected child → show Restore, undelete deleted child (because non-deleted items in the selection don't prevent Restore)

dstillman replied:
(Let's create tests for all of these.)

AbeJellinek replied:
See latest commits!

dstillman commented on chrome/content/zotero/zoteroPane.js:
```diff
@@ -1893,6 +1909,21 @@ var ZoteroPane = new function()
+
+
+	/**
+	 * Return whether every selected item is deleted.
```
Instead of having "Return" here (which would get real repetitive and is redundant with `@return`) and explaining the actual test (which could change), this can just be a more general description of what the function does. E.g., "Check whether all selected items can be restored from trash".

Also, nit: No need for periods at the end of comments

And same for the other function

dstillman reviewed (changes requested):
We should also show the appropriate item pane buttons based on these tests, so that they match the menu options. E.g., right now they're not visible when a child item of a deleted parent item is selected, even though the "Delete Permanently…" menu option is now available.

While we're at it, the menu option in the trash should really be "Delete Permanently…" (like the button) rather than "Delete Items…", which would also help explain why it's disabled if a parent item that isn't in the trash is selected.

AbeJellinek commented:
Updated restore behavior and replaced "Delete Items..." with "Delete Permanently...". I kept the `.multiple` variant around because there are languages in which that would be translated with a direct object pronoun that could be singular or plural.

dstillman commented:
> I kept the .multiple variant around because there are languages in which that would be translated with a direct object pronoun that could be singular or plural.

Do you have an example? I wouldn't think this is true. We [currently](https://github.com/zotero/zotero/pull/1535) use direct objects in all our menu options, but if we didn't, I wouldn't expect to have plural variants of every option that were just based on how many items you had selected. The Mac Finder doesn't use direct objects, and if your system is in French, it doesn't change every menu option based on the number of files selected — it just uses the infinitive:

<img width="284" alt="Screen Shot 2022-02-12 at 3 11 15 AM" src="https://user-images.githubusercontent.com/783082/153703210-860ae9f3-c2a3-41ac-b67f-7fa808e125f3.png">

<img width="350" alt="Screen Shot 2022-02-12 at 3 11 08 AM" src="https://user-images.githubusercontent.com/783082/153703214-96540fd9-9971-4478-9a2d-34c08874b0a2.png">

(In checking to see if this was universal, I found a [discussion](https://forum.wordreference.com/threads/en-edit-item-infinitive-imperative.2627307/) among French speakers about what the actual verb tense is in English menus, the answer being…¯\\\_(ツ)\_/¯.)

AbeJellinek commented:
>(In checking to see if this was universal, I found a [discussion](https://forum.wordreference.com/threads/en-edit-item-infinitive-imperative.2627307/) among French speakers about what the actual verb tense is in English menus, the answer being…¯\_(ツ)_/¯.)

Good to know that we can’t reuse button titles for window titles, too.

dstillman commented:
I'm not sure what verb forms for these are, but it doesn't look like they change:

<img width="220" alt="Screen Shot ٢٠٢٢-٠٢-١٣ at ١ ٣٥ ٣٠ ص" src="https://user-images.githubusercontent.com/783082/153742194-090ca60a-8c88-4704-864c-e3417b3021d0.png">

<img width="322" alt="Screen Shot 2022-02-13 at 1 35 38 AM" src="https://user-images.githubusercontent.com/783082/153742225-d8ef3a19-11dd-4357-9074-aa918f53ec5b.png">

I suspect this is pretty universal as a UI convention, so I think can remove the `multiple` variant.

(Even practically speaking it wouldn't really work, because you have to go out of your way to even see the key on Transifex, so everyone translating these identical English strings would just translate them as the same string.)

dstillman commented on chrome/content/zotero/zoteroPane.js:
```diff
@@ -1893,26 +1909,74 @@ var ZoteroPane = new function()
+					let children = [...parent.getNotes(true), ...parent.getAttachments(true)];
+					let noneSelected = !children.some(isSelected);
+					for (let child of Zotero.Items.get(children)) {
+						Zotero.debug('Child: ' + child.id, -1)
```
Remove

dstillman commented on chrome/content/zotero/zoteroPane.js:
```diff
@@ -1893,26 +1909,74 @@ var ZoteroPane = new function()
 			}
-		}.bind(this));
-	});
+			await Promise.all(savePromises);
```
We don't do this for object saves. It might be fine, or it might cause weird race conditions, but we don't design or test for it anywhere, and in any case it probably doesn't speed anything up. Can just use `await item.save()` within the loop, despite the eslint warning (which we should probably remove).

This makes more sense for things like network requests where a single request could take anywhere from 0 to many seconds and we don't want other requests to wait, but that's not an issue here.

dstillman commented:
Possible regression?

https://forums.zotero.org/discussion/94771/failed-to-restore-to-the-library
